javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/ExtensionClassLoader.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.tooling;
+
+import static io.opentelemetry.javaagent.tooling.ShadingRemapper.rule;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import net.bytebuddy.jar.asm.ClassReader;
+import net.bytebuddy.jar.asm.ClassWriter;
+import net.bytebuddy.jar.asm.commons.ClassRemapper;
+
+/**
+ * This classloader is used to load arbitrary extensions for Otel Java instrumentation agent. Such
+ * extensions may include SDK components (exporters or propagators) and additional instrumentations.
+ * They have to be isolated and shaded to reduce interference with the user application and to make
+ * it compatible with shaded SDK used by the agent.
+ */
+// TODO find a way to initialize logging before using this class
+public class ExtensionClassLoader extends URLClassLoader {
+
+  // We need to prefix the names to prevent the gradle shadowJar relocation rules from touching
+  // them. It's possible to do this by excluding this class from shading, but it may cause issue
+  // with transitive dependencies down the line.
+  private static final ShadingRemapper remapper =
+      new ShadingRemapper(
+          rule("#io.opentelemetry.api", "#io.opentelemetry.javaagent.shaded.io.opentelemetry.api"),
+          rule(
+              "#io.opentelemetry.context",
+              "#io.opentelemetry.javaagent.shaded.io.opentelemetry.context"),
+          rule(
+              "#io.opentelemetry.extension.aws",
+              "#io.opentelemetry.javaagent.shaded.io.opentelemetry.extension.aws"),
+          rule("#java.util.logging.Logger", "#io.opentelemetry.javaagent.bootstrap.PatchLogger"),
+          rule("#org.slf4j", "#io.opentelemetry.javaagent.slf4j"));
+
+  private final Manifest manifest;
+
+  public static ClassLoader getInstance(ClassLoader parent) {
+    // TODO add support for old deprecated properties, otel.exporter.jar and otel.initializer.jar
+    // TODO add support for system properties
+    URL extension = parseLocation(System.getenv("OTEL_JAVAAGENT_EXTENSIONS"));
+    if (extension != null) {
+      return new ExtensionClassLoader(extension, parent);
+    }
+    return null;
+  }
+
+  private static URL parseLocation(String name) {
+    if (name == null) {
+      return null;
+    }
+    try {
+      return new File(name).toURI().toURL();
+    } catch (MalformedURLException e) {
+      System.err.println("Filename could not be parsed: %s. Extension location is ignored");
+      e.printStackTrace();
+    }
+    return null;
+  }
+
+  public ExtensionClassLoader(URL url, ClassLoader parent) {
+    super(new URL[] {url}, parent);
+    this.manifest = getManifest(url);
```

Comment:
oh, I remember this, I'm not sure how to generalize this to more than one jar file under the constraint of using URLClassLoader

maybe we can pass URLs to URLClassLoader that have custom InputStreams that do the shading? If that can work somehow (big if) then URLClassLoader will take care of the class/package definitions for us and we don't need the manifest

Reply:
Hm... in fact we can do this, I think... I will give it a try.